rpk: add license commands (set and info)

r-vasquez opened the pull request:
## Cover letter

This enables the option to query admin api for license information and also set a new license.

Added 2 new commands, 

`rpk cluster license set`:

```
$ rpk cluster license set --path /path/to/license.dat --api-urls 0.0.0.0:9645 
Successfully uploaded license
```

and `rpk cluster license info`:

```
$ rpk cluster license info --api-urls 0.0.0.0:9645
LICENSE INFORMATION
====================
Organization:    Redpanda
Type:            Free trial
Expires:         40 days
Version:         0
```

or 

```
$ rpk cluster license info --api-urls 0.0.0.0:9645 --format json
{
  "format_version": 0,
  "org": "redpanda-testing",
  "type": "enterprise",
  "expires": 36489
}
```
Related issue: https://github.com/redpanda-data/core-internal/issues/20

## Release notes
* You can now manage your cluster licenses with `rpk cluster license`

emaxerrno commented:
@r-vasquez 
can we make the default `/etc/redpanda/redpanda.license` 


Also, a *requirement* is to display the organization. We cannot generate a valid license key without org. doesn't make sense. so the display `info` table should have the organization.

cc: @graphcareful - can you confirm that's the case. 
cc: @weeco

r-vasquez commented:
Update:

* Changed the way to display info (table to list)
* Changed property User -> Organization
* Default path to license is now /etc/redpanda/redpanda.license

jcsp commented:
This can probably proceed now that the core license code merged?

Would be good to add a couple of basic ducktape tests for the new commands for full end to end test.

graphcareful commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,49 @@
+	"github.com/spf13/cobra"
+)
+
+func newSetCommand(fs afero.Fs) *cobra.Command {
```
question: Was the idea that `--path` is optional and you can also pass it directly on the command line?

r-vasquez replied:
I think we agree on using `--path` and now I realize that passing it directly on the command line will make the license leak into shell history so it's better to use a file, wdyt?

graphcareful replied:
Yes IMO it being optional means you can have the benefits of both.

Good point about `--password`

The license key itself is never searched for in the env. In this PR when you see references to that, that is a unit test looking for a test license in the CI environment.

r-vasquez replied:
Ok, so --path should be the default and then allow another optional flag to pass in the license, wdyt of:

`rpk cluster license set --license <base64 encoded license>`

or `--raw-license`

twmb replied:
Why not `rpk cluster license set <base64 encoded>`?

r-vasquez replied:
Future proof and easier to deprecate: if we add another command (not likely) after `license set` it will be confusing to make the arg validation. Also, I was thinking that if the preferred way will be `--path` then it will be confusing to explain that the argument is optional.

graphcareful commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,59 @@
+	out.Section("LICENSE INFORMATION:")
+	licenseFormat := `Organization:    %v
+Type:            %v
+Expires:         %v
```
nit: Maybe we should write (days) or something to make this more obvious

graphcareful commented on tests/rptest/tests/rpk_cluster_test.py:
```diff
@@ -184,3 +187,61 @@ def test_cluster_down(self):
+        def condition():
+            output = self._rpk.license_info()
+
+            expected = '''LICENSE INFORMATION:
```
This is nice, just thinking the command could also support a `--json` option which would make testing easier

graphcareful replied:
Also I see we are not testing the number of days. I think that should be added to the test, since we know the exact date the license expires, just subtract that date from todays date and compare that to the value observed in `output['expires']`.

Here's an [example](https://github.com/redpanda-data/redpanda/blob/dev/tests/rptest/tests/cluster_features_test.py#L242-L243)

r-vasquez replied:
Done, thanks!

graphcareful commented on tests/rptest/tests/rpk_cluster_test.py:
```diff
@@ -184,3 +187,61 @@ def test_cluster_down(self):
+    def test_upload_cluster_license_error(self):
+        with expect_exception(RpkException,
+                              lambda e: "Internal Server Error" in str(e)):
+            license = os.environ.get("REDPANDA_SAMPLE_LICENSE", None)
```
This license env check stuff could be bundled in a util check

r-vasquez replied:
Done, let me know what do you think.

r-vasquez commented on tests/rptest/tests/rpk_cluster_test.py:
```diff
@@ -184,3 +187,61 @@ def test_cluster_down(self):
+        with tempfile.NamedTemporaryFile() as tf:
+            with open(tf.name, 'w') as f:
+                f.write(license)
+                f.seek(0)
```
Done, thanks!

twmb commented on src/go/rpk/pkg/api/admin/admin.go:
```diff
@@ -504,11 +504,16 @@ func (a *AdminAPI) sendAndReceive(
-		bs, err := json.Marshal(body)
-		if err != nil {
-			return nil, fmt.Errorf("unable to encode request body for %s %s: %w", method, url, err) // should not happen
+		// We might be passing io reader already as body, e.g: license file.
```
Can the doc on this function be updated with another line,
` // If the body is already an io.Reader, the reader is used directly without marshaling.`?

twmb commented on src/go/rpk/pkg/cli/cmd/cluster.go:
```diff
@@ -54,11 +55,13 @@ func NewClusterCommand(fs afero.Fs) *cobra.Command {
-	command.AddCommand(cluster.NewHealthOverviewCommand(fs))
+	command.AddCommand(
+		offsets,
+		config.NewConfigCommand(fs),
+		maintenance.NewMaintenanceCommand(fs),
+		cluster.NewHealthOverviewCommand(fs),
+		license.NewLicenseCommand(fs),
```
nit, but I try to keep these alphabetical

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,68 @@
+			out.MaybeDie(err, "unable to retrieve license info: %v", err)
+
+			if !info.Loaded {
+				out.Die("license is not loaded for this cluster")
```
Since this is a more proper exit, I think we may want a realer sentence here:
```
This cluster is missing a license.
```
?

r-vasquez replied:
I'll change that but lowercase and without a punctuation mark since we are sending it to stderr.

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,68 @@
+					printLicenseInfo(info.Properties)
+				}
+			} else {
+				out.Die("error retrieving loaded license information")
```
In what case are properties missing? Can we give a more meaningful message here?

cc @graphcareful

graphcareful replied:
Only in the case where a license is not loaded. If one exists but is expired it's metadata will be returned.

graphcareful replied:
So I agree, maybe a better message would be "no license loaded". Only display the error message if the admin server returns a 4xx or 5xx response

r-vasquez replied:
Changed for `no license loaded`

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,68 @@
+		},
+	}
+
+	command.Flags().BoolVar(&jsonFormat, "json", false, "Print license info in JSON format")
```
I did `-j` in kcl to output as json, but I think we should set outselves up for potentially other output formats down the line.

awscli uses `--output`, gcloud uses `--format`, and I think what's clearest is `--output-format`. Does any other CLI do something similar? If no, I think we should lean `--output-format` rather than `--json`. Down the line, I expect we will expand formatting to other commands.

r-vasquez replied:
Went with `--format` since `rpk topic consume` already uses `--format`

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,68 @@
+}
+
+func printLicenseInfo(p admin.LicenseProperties) {
+	out.Section("LICENSE INFORMATION:")
```
no colon, since a section is already underlined with a bunch of ='s, the colon seems a bit redundant.

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,49 @@
+Command will require you to provide a path to the license file:
+
+    rpk cluster license set --path /home/organization/redpanda.license
+
```
this desc will need updating pending the inline base64 arg discussion above

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,49 @@
+			defer file.Close()
+
+			err = cl.SetLicense(cmd.Context(), file)
+			out.MaybeDie(err, "unable to set license :%q, error: %v", licPath, err)
```
` :%q` => `%q`

r-vasquez commented:
Update:

- Changed `--json` to `--format`.
- Added more ducktape tests.
- PR comments.

twmb commented on tests/rptest/clients/rpk.py:
```diff
@@ -713,3 +713,26 @@ def cluster_metadata_id(self):
+        cmd = [
+            self._rpk_binary(), "--api-urls",
+            self._admin_host(), "cluster", "license", "info", "--format",
+            "json"
```
why is `"json"` all alone on this line? Is there some formatter?

r-vasquez replied:
yapf formatter :D

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,71 @@
+
+		Run: func(cmd *cobra.Command, args []string) {
+			if licPath != "" && len(args) > 0 {
+				out.Die("inline license can't be passed if flag '--path' is set")
```
`can't` => `cannot` (I try to avoid contractions in CLI)

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,71 @@
+				out.Die("inline license can't be passed if flag '--path' is set")
+			}
+			if licPath == "" && len(args) == 0 {
+				fmt.Println("Neither license file nor inline license was provided, looking for license file in '/etc/redpanda/redpanda.license'")
```
How about `looking for license file in` => `checking`, to make the line shorter?
also suffix with period

twmb replied:
Actually, why not make `/etc/redpanda/redpanda.license` the default value for the flag, and if both are empty here, exit with an error?

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,71 @@
+			err = cl.SetLicense(cmd.Context(), r)
+			out.MaybeDie(err, "unable to set license: %v", err)
+
+			fmt.Println("Successfully uploaded license")
```
suffix with period

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,71 @@
+
+Or inline the license string:
+
+    rpk cluster license set <license string>
```
`<base64 license string>`?

graphcareful reviewed (commented):
Nice updates with TLS, should we test it , or leave it up to the fact that it calls existing TLS constructs that are already tested?

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,68 @@
+			if info.Properties != (admin.LicenseProperties{}) {
+				if format == "json" {
+					props, err := json.MarshalIndent(info.Properties, "", "  ")
+					out.MaybeDie(err, "unable to print license information as json: %v", err)
+					fmt.Printf("%s\n", props)
```
fwiw this _cannot_ error per the input type. There's a linter that actually checks this so that you can drop the `err` arg, but it doesn't catch that Encode can fail for other reasons

tldr; I'd drop the fmt.Printf. The MaybeDie can be kept for completeness sake

twmb replied:
(misread this line, deleting comment)

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/info.go:
```diff
@@ -0,0 +1,68 @@
+		},
+	}
+
+	command.Flags().StringVar(&format, "format", "text", "Output format. Supported formats: text, json")
```
`Output format (text, json)`

I think that's closer to how our other flags format printing available options, but I'm not 100% sure. I know we usually avoid sentences within flags.

twmb commented on src/go/rpk/pkg/cli/cmd/cluster/license/set.go:
```diff
@@ -0,0 +1,71 @@
+    rpk cluster license set <license string>
+
+If neither are present, rpk will look for the license in the
+default location '/etc/redpanda/redpanda.license'
```
Can this end with a period?

r-vasquez commented:
Last build failure https://buildkite.com/redpanda/redpanda/builds/12554#0181fe56-1eca-4f9e-9ab4-a2261e3c50aa

Related to: https://github.com/redpanda-data/redpanda/issues/5466
